### PR TITLE
Dns over http2 5773 v18.1

### PR DESCRIPTION
**Pull request by catenacyber**
Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
https://redmine.openinfosecfoundation.org/issues/5773

Describe changes:
- analyze DNS over HTTP2

SV_BRANCH=https://github.com/OISF/suricata-verify/pull/1980

https://github.com/OISF/suricata/pull/11498 with needed rebase

Should there be a squash up of commits ?

@jasonish still same question : here for a DOH2 tx, we log a bidirectional HTTP2 transaction, and then if any, a DNS transaction, preferring the answer... What do you think about it ? This allows to keep the same format as for regular dns.
Another option would be to log two doh2 events : one for the DNS request and one for the DNS answer, with HTTP2 getting logged twice... not sure how it would work out for alerts...

### CODE REVIEW

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 21626

File: rust/src/http2/http2.rs [hunk 1]
```diff
@@ -301,13 +357,23 @@ impl HTTP2Transaction {
                 &xid,
             );
         };
+        if unsafe { ALPROTO_DOH2 } != ALPROTO_UNKNOWN {
+            // we store DNS response, and process it when complete
+            if let Some(doh) = &mut self.doh {
+                if doh.is_doh_data[dir.index()] && doh.data_buf[dir.index()].len() < 0xFFFF {
```
Comment (victorjulien) [inline rust/src/http2/http2.rs:363]: should there be an event if we exceed this?

File: rust/src/http2/http2.rs [hunk 2]
```diff
@@ -1404,4 +1543,22 @@ pub unsafe extern "C" fn rs_http2_register_parser() {
     } else {
         SCLogNotice!("Protocol detector and parser disabled for HTTP2.");
     }
+
+    // doh2 is just http2 wrapped in another name
+    parser.name = b"doh2\0".as_ptr() as *const std::os::raw::c_char;
```
Comment (victorjulien) [inline rust/src/http2/http2.rs:1548]: should there be a check that http2 is enabled?

File: suricata.yaml.in [hunk 3]
```diff
@@ -948,6 +950,8 @@ app-layer:
     ssh:
       enabled: yes
       #hassh: yes
+    doh2:
```
Comment (victorjulien) [inline suricata.yaml.in:953]: comment to explain it is dns over http2, depends on http2 parsing

File: src/output-json-dns.c [hunk 4]
```diff
@@ -254,6 +254,89 @@ bool AlertJsonDns(void *txptr, JsonBuilder *js)
...
+        jb_restore_mark(jb, &mark);
+    }
+
+    void *tx_dns = DetectGetInnerTx(txptr, ALPROTO_DOH2, ALPROTO_DNS, STREAM_TOCLIENT);
+    if (tx_dns == NULL) {
+        tx_dns = DetectGetInnerTx(txptr, ALPROTO_DOH2, ALPROTO_DNS, STREAM_TOSERVER);
+    }
+    bool r2 = false;
+    if (tx_dns) {
+        // mix of JsonDnsLogger
```
Comment (victorjulien) [inline src/output-json-dns.c:310]: don't understand this comment

File: src/detect.c [hunk 5]
```diff
@@ -1050,6 +1050,24 @@ DetectRunTxSortHelper(const void *a, const void *b)
 #define TRACE_SID_TXS(sid,txs,...)
 #endif
 
+// Get inner transaction for engine
+void *DetectGetInnerTx(void *tx_ptr, AppProto alproto, AppProto engine_alproto, uint8_t flow_flags)
```
Comment (victorjulien) [inline src/detect.c:1054]: could be inline since its called a lot?

**Review by victorjulien — APPROVED** [review]
Some nits and minor questions inline.

**Comment by victorjulien** [comment]
Wrt the logging, I think we can sort this out after merge.

File: src/app-layer-protos.c [hunk 6]
```diff
@@ -62,6 +62,7 @@ const AppProtoStringTuple AppProtoStrings[ALPROTO_MAX] = {
     { ALPROTO_TELNET, "telnet" },
     { ALPROTO_WEBSOCKET, "websocket" },
     { ALPROTO_LDAP, "ldap" },
+    { ALPROTO_DOH2, "doh2" },
```
Comment (victorjulien) [inline src/app-layer-protos.c:65]: is this an generally used name for DNS over HTTP/2? Or are ppl just using doh? could perhaps use `dns-http2` or similar

File: src/util-profiling.h [hunk 7]
```diff
@@ -170,13 +170,12 @@ PktProfiling *SCProfilePacketStart(void);
         (dp)->alproto = (id);                                       \
     }
 
-#define PACKET_PROFILING_APP_END(dp, id)                            \
-    if (profiling_packets_enabled) {                                \
-        BUG_ON((id) != (dp)->alproto);                              \
```
Comment (inashivb) [inline src/util-profiling.h:175]: what about the other protocols where this should still hold?

File: rust/src/http2/http2.rs [hunk 8]
```diff
@@ -151,11 +164,7 @@ pub struct HTTP2Transaction {
     pub req_line: Vec<u8>,
     pub resp_line: Vec<u8>,
 
-    is_doh_data: [bool; 2],
-    // dns response buffer
-    pub doh_data_buf: [Vec<u8>; 2],
-    pub dns_request_tx: Option<DNSTransaction>,
-    pub dns_response_tx: Option<DNSTransaction>,
+    pub doh: Option<DohHttp2Tx>,
```
Comment (inashivb) [inline rust/src/http2/http2.rs:167]: noice :)

**Comment by victorjulien** [comment]
Merged in #11536, thanks!

**Pull request closed without merging**